manifest build trips over "developcopy" env paths embedded in symlinks

**Describe the bug:**

During a manifest build the `$FLOX_ENV` variable is set to point to a _copy_ of the "develop" environment, referred to here by the name "developcopy". At the conclusion of the build references to the "developcopy" environment path are then replaced with the "build wrapper" environment path, and it does this by rewriting strings in a stream of bytes created by `gnutar` on one end and consumed by `gnutar` on the other.

This rewriting is very effective, replacing paths within symbolic link values just as easily as within files. The trouble is that rewriting the value of a symlink in this way triggers a gnutar error and breaks the build.

**Steps to reproduce:**
1. Create flox environment with following manifest:
```
version = 1
[build.foo]
command = """
  mkdir -p $out
  ln -s $FLOX_ENV/bin $out/bin
"""
```
2. Invoke `flox build`
3. Observe failure

Result:
```
flox [limeytexan/default] Michaels-MBP% flox build
Rendering foo build script to /tmp/8a65d2c3/foo/build.bash
Building foo-0.0.0 in local mode
00:00:00.003616 + mkdir -p /tmp/store_d5ba076d76292a45ead13b256e75e2e5-foo-0.0.0
00:00:00.006386 + ln -s /nix/store/q8j5632n0qrb3agc7hvk1glcs2hzxjqf-developcopy-build-foo/bin /tmp/store_d5ba076d76292a45ead13b256e75e2e5-foo-0.0.0/bin
/nix/store/rphdiw3w4s7cy16z1y0slryjgaw3xal6-gnutar-1.35/bin/tar: Skipping to next header
/nix/store/rphdiw3w4s7cy16z1y0slryjgaw3xal6-gnutar-1.35/bin/tar: Exiting with failure status due to previous errors
make: *** [/nix/store/g0c14470ixzqrd7q35q61wipy8s3w9rr-package-builder-1.0.0/libexec/flox-build.mk:718: /tmp/store_d5ba076d76292a45ead13b256e75e2e5-foo-0.0.0] Error 2
❌ ERROR: Build failed with status: exit status: 2
```

Expected:

Not to fail.

**Screenshots/terminal output (if relevant):**

**Flox Version (run `flox --version` if possible):**
```
flox [limeytexan/default] Michaels-MBP% flox --version
1.5.0-g251a45e
```

**`uname -a` output:**

--- a/cli/flox-rust-sdk/src/providers/build.rs
+++ b/cli/flox-rust-sdk/src/providers/build.rs
@@ -2601,6 +2601,54 @@ mod tests {
         let (flox, tmpdir) = flox_instance();
         assert_manifest_build_succeeds(GENERATED_DATA.join("build/hello"), "hello", &flox, tmpdir);
     }
+
+    async fn build_symlinks_can_refer_to_flox_env(sandbox: &str) {
+        let package_name = String::from("foo");
+        let manifest = formatdoc! {r#"
+            version = 1
+            [install]
+            hello.pkg-path = "hello"
+
+            [build.{package_name}]
+            sandbox = "{sandbox}"
+            command = """
+                mkdir -p $out/bin
+                ln -s $FLOX_ENV/bin/hello $out/bin
+            """
+        "#};
+
+        let (mut flox, _temp_dir_handle) = flox_instance();
+        let mut env = new_path_environment(&flox, &manifest);
+        let env_path = env.parent_path().unwrap();
+        let _ = GitCommandProvider::init(&env_path, false).unwrap();
+        flox.catalog_client =
+            catalog_replay_client(GENERATED_DATA.join("resolve/hello.yaml")).await;
+        assert_build_status(&flox, &mut env, &package_name, None, true);
+
+        let result_path = result_dir(&env_path, &package_name)
+            .join("bin")
+            .join("hello");
+        let output = Command::new(&result_path).output().unwrap();
+        assert!(
+            output.status.success(),
+            "should execute successfully, stderr: {:?}",
+            String::from_utf8_lossy(&output.stderr)
+        );
+        assert_eq!(
+            String::from_utf8_lossy(&output.stdout).trim_end(),
+            "Hello, world!"
+        );
+    }
+
+    #[tokio::test(flavor = "multi_thread")]
+    async fn build_symlinks_can_refer_to_flox_env_sandbox_pure() {
+        build_symlinks_can_refer_to_flox_env("pure").await;
+    }
+
+    #[tokio::test(flavor = "multi_thread")]
+    async fn build_symlinks_can_refer_to_flox_env_sandbox_off() {
+        build_symlinks_can_refer_to_flox_env("off").await;
+    }
 }
 
 #[cfg(test)]

--- a/package-builder/build-manifest.nix
+++ b/package-builder/build-manifest.nix
@@ -90,6 +90,7 @@ pkgs.runCommandNoCC name
     nativeBuildInputs =
       with pkgs;
       [
+        cpio
         findutils
         gnutar
         gnused
@@ -129,9 +130,11 @@ pkgs.runCommandNoCC name
             # to the output directory, rewriting path references as we go.
             if [ -d ${install-prefix-contents} ]; then
               mkdir $out
-              tar -C ${install-prefix-contents} -c --mode=u+w -f - . | \
+              ( cd ${install-prefix-contents} && find . -print0 | \
+                cpio --null --create --format newc ) | \
                 sed --binary "s%${install-prefix}%$out%g" | \
-                tar -C $out -xf -
+                ( cd $out && cpio --extract --make-directories --preserve-modification-time \
+                  --unconditional --no-absolute-filenames --quiet && chmod -R u+w . )
             else
               cp ${install-prefix-contents} $out
               sed --binary "s%${install-prefix}%$out%g" $out
@@ -227,9 +230,11 @@ pkgs.runCommandNoCC name
             mv "$out" "$TMPDIR/$bn"
             if [ -d "$TMPDIR/$bn" ]; then
               mkdir "$out"
-              tar -C "$TMPDIR/$bn" -c --mode=u+w -f - . | \
+              ( cd "$TMPDIR/$bn" && find . -print0 | \
+                cpio --null --create --format newc ) | \
                 sed --binary "s%${develop-copy-env-package}%${build-wrapper-env-package}%g" | \
-                tar -C "$out" -xf -
+                ( cd $out && cpio --extract --make-directories --preserve-modification-time \
+                  --unconditional --no-absolute-filenames --quiet && chmod -R u+w . )
             else
               sed --binary "s%${develop-copy-env-package}%${build-wrapper-env-package}%g" < "$TMPDIR/$bn" > "$out"
             fi

--- a/package-builder/flox-build.mk
+++ b/package-builder/flox-build.mk
@@ -31,7 +31,9 @@ comma := ,
 # Substitute Nix store paths for packages required by this Makefile.
 __bashInteractive := @bashInteractive@
 __coreutils := @coreutils@
+__cpio := @cpio@
 __daemonize := @daemonize@
+__findutils := @findutils@
 __gitMinimal := @gitMinimal@
 __gnugrep := @gnugrep@
 __gnused := @gnused@
@@ -48,10 +50,13 @@ __t3 := @t3@
 __package_bin = $(if $(filter @%@,$(1)),$(2),$(1)/bin/$(2))
 _bash := $(call __package_bin,$(__bashInteractive),bash)
 _cat := $(call __package_bin,$(__coreutils),cat)
+_chmod := $(call __package_bin,$(__coreutils),chmod)
 _cp := $(call __package_bin,$(__coreutils),cp)
+_cpio := $(call __package_bin,$(__cpio),cpio)
 _cut := $(call __package_bin,$(__coreutils),cut)
 _daemonize := $(call __package_bin,$(__daemonize),daemonize)
 _env := $(call __package_bin,$(__coreutils),env)
+_find := $(call __package_bin,$(__findutils),find)
 _git := $(call __package_bin,$(__gitMinimal),git)
 _grep := $(call __package_bin,$(__gnugrep),grep)
 _head := $(call __package_bin,$(__coreutils),head)
@@ -432,15 +437,17 @@ define BUILD_local_template =
 	    $(_t3) $($(_pvarname)_logfile) -- $(_bash) -e $$<
 	@#
 	@# Finally, rewrite references to temporary build wrapper in "out",
-	@# making sure to render the substituted output in the same location
+	@# making sure to return the substituted output to the same location
 	@# in which it was built.
 	@#
 	$(_VV_) if [ -d "$($(_pvarname)_out)" ]; then \
 	  $(_mkdir) -p $($(_pvarname)_out).new && \
 	  set -o pipefail && \
-	    $(_tar) -C $($(_pvarname)_out) -c --mode=u+w -f - . | \
+	    ( cd $($(_pvarname)_out) && $(_find) . -print0 | \
+	      $(_cpio) --null --create --format newc ) | \
 	    $(_sed) --binary "s%$$($(_pvarname)_develop_copy_env)%$$($(_pvarname)_build_wrapper_env)%g" | \
-	    $(_tar) -C $($(_pvarname)_out).new -xf - && \
+	    ( cd $($(_pvarname)_out).new && $(_cpio) --extract --make-directories --preserve-modification-time \
+	      --unconditional --no-absolute-filenames --quiet && $(_chmod) -R u+w . ) && \
 	  $(_rm) -rf $($(_pvarname)_out) && \
 	  $(_mv) $($(_pvarname)_out).new $($(_pvarname)_out); \
 	fi

--- a/pkgs/flox-package-builder/default.nix
+++ b/pkgs/flox-package-builder/default.nix
@@ -1,6 +1,7 @@
 {
   bashInteractive,
   coreutils,
+  cpio,
   daemonize,
   findutils,
   getopt,
@@ -28,6 +29,7 @@ stdenv.mkDerivation {
     for i in build-manifest.nix flox-build.mk validate-build.bash; do
       bashInteractive=${bashInteractive} \
       coreutils=${coreutils} \
+      cpio=${cpio} \
       daemonize=${daemonize} \
       findutils=${findutils} \
       getopt=${getopt} \

